bottlecap/src/secrets/decrypt.rs

```diff
@@ -43,23 +56,75 @@ pub fn resolve_secrets(config: Config) -> Result<Config> {
     }
 }
 
+struct RequestArgs<'a> {
+    service: String,
+    body: &'a Value,
+    time: DateTime<Utc>,
+    x_amz_target: String,
+}
+
 struct AwsConfig {
     region: String,
     aws_access_key_id: String,
     aws_secret_access_key: String,
     aws_session_token: String,
+    function_name: String,
 }
 
-fn manual_decrypt(secret_arn: String, aws_config: AwsConfig) -> Result<String> {
+fn decrypt_aws_kms(client: Client, kms_key: String, aws_config: AwsConfig) -> Result<String> {
+    let json_body = &serde_json::json!({
+        "CiphertextBlob": kms_key,
+        "encryptionContext": { "LambdaFunctionName": aws_config.function_name }}
+    );
+
+    let headers = build_get_secret_signed_headers(
+        &aws_config,
+        RequestArgs {
+            service: "kms".to_string(),
+            body: json_body,
+            time: Utc::now(),
+            x_amz_target: "TrentService.Decrypt".to_string(),
+        },
+    );
+
+    let v = request(json_body, headers, client);
+
+    return if let Some(secret_string_b64) = v["Plaintext"].as_str() {
+        let secret_string = String::from_utf8(
+            BASE64_STANDARD
+                .decode(secret_string_b64)
+                .expect("Failed to decode base64"),
+        )
+        .expect("Failed to convert to string");
+        Ok(secret_string)
+    } else {
+        Err(Error::new(std::io::ErrorKind::InvalidData, v.to_string()))
+    };
+}
+
+fn decrypt_aws_sm(client: Client, secret_arn: String, aws_config: AwsConfig) -> Result<String> {
     let json_body = &serde_json::json!({ "SecretId": secret_arn});
 
-    let headers = build_get_secret_signed_headers(&aws_config, json_body, Utc::now());
+    let headers = build_get_secret_signed_headers(
+        &aws_config,
+        RequestArgs {
+            service: "secretsmanager".to_string(),
+            body: json_body,
+            time: Utc::now(),
+            x_amz_target: "secretsmanager.GetSecretValue".to_string(),
+        },
+    );
+
+    let v = request(json_body, headers, client);
 
-    let client = reqwest::blocking::Client::builder()
-        .use_rustls_tls()
-        .build()
-        .expect("Failed to create reqwest client for aws decrypt");
+    return if let Some(secret_string) = v["SecretString"].as_str() {
```

Comment:
Return is not needed if last line is already there, same suggestion as before, maybe we should do a Result<String, ...> where we can bubble up errors?

Reply:
I'm not a big fan of omitting return, but I will stick with the idiom!